integrations/nifi/nifi-nar-bundles/nifi-standard-services/nifi-spark-service-bundle/nifi-spark-service/src/main/java/com/thinkbiganalytics/nifi/v2/spark/SparkJobserverService.java

```diff
@@ -0,0 +1,353 @@
+                } finally {
+                    synchronized (sparkContextsStates) {
+                        if (contextRunning) {
+                            sparkContextsStates.get(contextName).setStateAsRunning();
+                            sparkContextsStates.get(contextName).setTimeoutSeconds(contextTimeout);
+                        } else {
+                            sparkContextsStates.remove(contextName);
+                        }
+                        sparkContextsStates.notifyAll();
+                    }
+                }
+            }
+            return contextRunning;
+        } else {
+            return true;
+        }
+    }
+
+    @Override
+    public boolean deleteContext(String contextName) {
+
+        Boolean contextDeleted = !checkIfContextExists(contextName);
+
+        if (!contextDeleted) {
+            try {
+                synchronized (sparkContextsStates) {
+                    contextDeleted = client.deleteContext(contextName);
+                    if (contextDeleted){
+                        sparkContextsStates.remove(contextName);
+                        getLogger().info(String.format("Deleted context %s from Spark Jobserver %s", contextName, jobServerUrl));
+                    }
+                }
+            } catch (Exception ex) {
+                getLogger().error(ex.getMessage());
+            }
+        }
+        return contextDeleted;
+    }
+
+    @Override
+    public boolean executeSparkContextJob(String appName, String classPath, String contextName, String args, boolean async) {
+
+        String id = contextName + System.nanoTime();
+
+        synchronized (sparkContextsStates) {
+            if (sparkContextsStates.containsKey(contextName)) {
+                sparkContextsStates.get(contextName).addExecutionLock(id);
+            }
+        }
+
+        SparkJobResult jobResult = null;
+        Boolean success = false;
+
+        try {
+            getLogger().info(String.format("Executing Spark App %s %s on context %s with args %s on Spark Jobserver %s", appName, classPath, contextName, args, jobServerUrl));
+
+            Map<String, String> params = new HashMap<>();
+
+            params.put(ISparkJobServerClientConstants.PARAM_APP_NAME, appName);
+            params.put(ISparkJobServerClientConstants.PARAM_CLASS_PATH, classPath);
+            params.put(ISparkJobServerClientConstants.PARAM_CONTEXT, contextName);
+
+            if (async) {
+                params.put(ISparkJobServerClientConstants.PARAM_SYNC, "false");
+            } else {
+                params.put(ISparkJobServerClientConstants.PARAM_SYNC, "true");
+                params.put(ISparkJobServerClientConstants.PARAM_TIMEOUT, syncTimeout);
+            }
+
+            jobResult = client.startJob(args, params);
+
+            getLogger().info(String.format("Executed %s %s on context %s on Spark Jobserver %s", appName, classPath, contextName, jobServerUrl));
```

Comment:
Please allow `info` to format the message rather than using `String.format`. This is more performant if the INFO logging is not enabled.